File: app/src/main/java/com/va/android/task/implementation/java/util/SimpleCountingIdlingResource.java
package com.va.android.task.implementation.java.util;

import java.util.concurrent.atomic.AtomicInteger;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.test.espresso.IdlingResource;

/**
 * An simple counter implementation of {@link IdlingResource} that determines idleness by
 * maintaining an internal counter. When the counter is 0 - it is considered to be idle, when it is
 * non-zero it is not idle.
 *
 * This class can then be used to wrap up operations that while in progress should block tests from
 * accessing the UI.
 */
public final class SimpleCountingIdlingResource implements IdlingResource {

    @Nullable
    private volatile ResourceCallback mCallback;
    private final AtomicInteger mCounter = new AtomicInteger(0);
    private final String mName;

    public SimpleCountingIdlingResource(@NonNull String name) {
        mName = name;
    }

    @Override
    public String getName() {
        return mName;
    }

    @Override
    public boolean isIdleNow() {
        return mCounter.get() == 0;
    }

    @Override
    public void registerIdleTransitionCallback(ResourceCallback callback) {
        mCallback = callback;
    }

    public void increment() {
        mCounter.getAndIncrement();
    }

    public void decrement() {
        int value = mCounter.decrementAndGet();
        if (value == 0 && mCallback != null) {
            mCallback.onTransitionToIdle();
        }
    }
}

File: app/src/main/java/com/example/android/scheduler/implementation/kotlin/util/SimpleCountingIdlingResource.kt
package com.example.android.scheduler.implementation.kotlin.util

import androidx.test.espresso.IdlingResource
import androidx.test.espresso.IdlingResource.ResourceCallback

import java.util.concurrent.atomic.AtomicInteger

/**
 * An simple counter implementation of [IdlingResource] that determines idleness by
 * maintaining an internal counter. When the counter is 0 - it is considered to be idle, when it is
 * non-zero it is not idle.
 *
 * This class can then be used to wrap up operations that while in progress should block tests from
 * accessing the UI.
 */
class SimpleCountingIdlingResource(private val name: String) : IdlingResource {

    @Volatile
    private var callback: ResourceCallback? = null
    private val counter = AtomicInteger(0)

    override fun getName(): String = name

    override fun isIdleNow(): Boolean = counter.get() == 0

    override fun registerIdleTransitionCallback(callback: ResourceCallback?) {
        this.callback = callback
    }

    fun increment() {
        counter.getAndIncrement()
    }

    fun decrement() {
        val count = counter.decrementAndGet()
        if (count == 0) {
            callback?.onTransitionToIdle()
        }
    }
}